src/algebra/non_associative/basic.lean

```diff
@@ -0,0 +1,155 @@
+/-
+Copyright (c) 2021 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import algebra.algebra.basic
+
+/-!
+# Non-associative, non-unital algebras
+
+This files defines non-associative, non-unital algebras together with their morphisms and basic
+lemmas to make them usable.
+
+## Main definitions
+
+  * `na_semiring`
+  * `na_algebra`
+  * `na_algebra_hom`
+
+## Tags
+
+non-associative, non-unital, algebra
+-/
+
+universes u₁ u₂ u₃ u₄
+
+variables (R : Type u₁) (A : Type u₂) (B : Type u₃) (C : Type u₄) [semiring R]
+
+set_option old_structure_cmd true
+
+/-- A not-necessarily-unital, not-necessarily-associative semiring, or "non-associative ring" for
+short. -/
+class na_semiring extends add_comm_monoid A, distrib A, mul_zero_class A
+
+-- https://github.com/leanprover-community/lean/issues/285
+set_option old_structure_cmd false
+
+/-- A not-necessarily-unital, not-necessarily-associative algebra, or "non-associative algebra" for
+short. -/
+@[protect_proj]
+class na_algebra [na_semiring A] extends semimodule R A :=
+(smul_mul_assoc' : ∀ (t : R) (a b : A), (t • a) * b = t • (a * b))
+(mul_smul_comm' : ∀ (t : R) (a b : A), a * (t • b) = t • (a * b))
+
+@[priority 200] -- see Note [lower instance priority]
+instance semiring.to_na_semiring [semiring A] : na_semiring A := { ..(infer_instance : semiring A) }
+
+@[priority 200] -- see Note [lower instance priority]
+instance algebra.to_na_algebra (R : Type u₁) [comm_semiring R] [semiring A] [algebra R A] :
+  na_algebra R A :=
+{ smul_mul_assoc' := algebra.smul_mul_assoc,
+  mul_smul_comm'  := algebra.mul_smul_comm,
+  ..algebra.to_semimodule }
+
+namespace na_algebra
+
+variables {R A} [na_semiring A] [na_algebra R A]
+
+@[simp] lemma smul_mul_assoc (t : R) (a b : A) :
+  (t • a) * b = t • (a * b) :=
+na_algebra.smul_mul_assoc' t a b
+
+@[simp] lemma mul_smul_comm (t : R) (a b : A) :
+  a * (t • b) = t • (a * b) :=
+na_algebra.mul_smul_comm' t a b
+
+/-- The algebra multiplication as a bilinear map. -/
+def mul_as_bilinear {R : Type u₁} {A : Type u₂} [comm_semiring R] [na_semiring A] [na_algebra R A] :
+  A →ₗ[R] A →ₗ[R] A :=
+linear_map.mk₂ R (*) add_mul smul_mul_assoc mul_add mul_smul_comm
+
+end na_algebra
+
+variables [na_semiring A] [na_algebra R A]
+variables [na_semiring B] [na_algebra R B]
+variables [na_semiring C] [na_algebra R C]
+
+set_option old_structure_cmd true
+
+/-- A morphism of non-associative algebras. -/
+structure na_algebra_hom extends A →ₗ[R] B, mul_hom A B
```

Comment:
Strictly this is just about non-unital algebras, right? You'd use the same type whether or not the algebra is associative.

Reply:
Yes, quite right. At the very least, I should reword the comment. I'll wait and see what others think before I start updating names though since I'm fairly sure that `na_algebra` won't survive!

Reply:
Would `nassoc_algebra` be too long?

Reply:
I'd probably lean towards `pre_algebra`, since the only reason we don't want to use this as `algebra` is definitional - the axiomatic content is equivalent in all the places that `algebra` is used today.

Reply:
I don't like the idea of introducing new terminology that everyone has to look up. According to Wikipedia, non-associative or distributive are used: https://en.m.wikipedia.org/wiki/Non-associative_algebra

Reply:
@Julian-Kuelshammer, my point is that for "non-associativity", we may as well just redefine `algebra` to require `na_semiring` instead of `semiring`. Existing lemmas that assume `semiring` will keep working.

It's only non-unitarity that results in needing a separate class.

Reply:
I think it is quite common in mathematics to study non-associative algebras over commutative associative (semi)rings. That's not too far from my own area of research. I have not seen algebras over non-associative rings (but that might be a fault in my perspective).

Reply:
@Julian-Kuelshammer I don't think @eric-wieser is suggesting having non-associative coefficients (I too find this very exotic, though perhaps hard core groups theorists study Lie ring modules?).

Ignoring the detail that I have proposed allowing non-commutative coefficients for non-assoc algebras, I think Eric is pointing out that we don't really need to have both `algebra` and `na_algebra` classes since they could have the same definition and differ only in whether they consume a`na_semiring` or a `semiring` class.

I wholeheartedly agree with this point but I worry that the required refactor (changing the extant definition of `algebra`) would stall my efforts to get non-associative algebras off the ground. I thus advocate the approach proposed here: get started with `na_algebra` and refactor later.

Oh and on naming, I like `nassoc_algebra` best.

Reply:
In the end I have opted for the following names:

- `non_unital_non_assoc_semiring`
- `non_unital_semiring`
- `non_assoc_semiring`
- `non_unital_non_assoc_algebra`
- `non_unital_non_assoc_algebra_hom`